design/cdk-bootstrap.md

```diff
@@ -0,0 +1,444 @@
+  that should be attached to the **CFN deployment role**
+  instead of choosing them from a menu interactively.
+
+* `--qualifier`: an optional qualifier added to the physical names of the bootstrap resources
+  in order to protect against name collisions, especially for things like S3 buckets
+  (which are globally unique in a given AWS partition).
+
+## Updating bootstrap resources
+
+Because we already have a bootstrap solution in place,
+and it's possible we will need to add more bootstrap resources as time goes by,
+we should have a mechanism in place for migrating,
+and giving meaningful errors if the bootstrapping has not been done for an environment that needs it.
+
+I don't think invoking the full `cdk-bootstrap` tool on every deploy is a good idea, though;
+I worry that calculating a full diff of actual versus desired resource state might impact the performance of commands like
+`deploy` too negatively.
+
+My proposal is to have an export on the bootstrap stack,
+called `BootstrapVersion`, that will simply contain a number.
+We will start with the bootstrap template setting that export to the value `1`.
+With time, as we change the bootstrap template,
+we will increment the version export number.
+
+In the `cdk` commands,
+we can add a CLI option that will perform a 'bootstrap version check'
+before doing any operations.
+It will call the `DescribeStack` CFN API,
+and get the value of the `BootstrapVersion` export.
+Depending on the value retrieved, it will then:
+
+* If no such stack was found, that means bootstrapping was not performed for this environment.
+  Fail with the appropriate message.
+
+* If the stack was found, but it didn't have an export called `BootstrapVersion`,
+  that means the bootstrap stack is of an older version than the used CLI version,
+  and needs to be updated.
+  Fail with the appropriate message.
+
+* If the export is the same as the `BOOTSTRAP_VERSION` constant in the current CLI,
+  everything is fine - nothing to do.
+
+* If the export value is smaller than the `BOOTSTRAP_VERSION` constant in the current CLI,
+  that means the bootstrap stack is of an older version than the used CLI version,
+  and needs to be updated.
+  Fail with the appropriate message.
+
+* If the export value is larger than the `BOOTSTRAP_VERSION` constant in the current CLI,
+  that means the bootstrap stack is actually from a later version than the used CLI version.
+  In this case, I think it's correct to proceed with carrying out the operation;
+  perhaps print a warning that the user should consider updating their CLI version
+  if they encounter any errors.
+
+## Bootstrap template
+
+Here is the JSON of the bootstrap CloudFormation template:
+
+```json
+{
+  "Parameters": {
+    "ResourceNamePrefix": {
+      "Description": "The prefix to use for all of the physical names of the bootstrap resources",
+      "Default": "aws-cdk-bootstrap-hnb659fds",
+      "Type": "String"
+    },
+    "StagingBucketName": {
```

Comment:
Remove these please. All names should be fixed.

Reply:
Removed.